File: from-books/hash-table-exercises/language/javascript/exercise-3.ts
/*
 * Write a function that acceps a string that contains all the letters of the alphabet
 * except one and returns the missing letter. For example, the string "the quick brown box
 * jumps over a lazy dog" contains all the letters of the alphabet except the letter, "f".
 * The function should have a time complexity of O(n).
 */
function generateAlphabet(): { [index: string]: number } {
  const alphabetMap = {};
  for (let i = 0; i < 26; ++i) {
    let letter = String.fromCharCode(97 + i);
    alphabetMap[letter] = i + 1;
  }

  return alphabetMap;
}

function missingLetter(sentence: string): string {
  const alphabet: { [index: string]: number } = generateAlphabet();
  const sentenceMap: { [index: string]: number } = {};

  for (const letter of sentence) {
    if (!sentenceMap[letter]) {
      sentenceMap[letter] = 1;
    }
  }

  let missing = "";
  for (const letter in alphabet) {
    if (!sentenceMap[letter]) {
      missing = letter;
      break;
    }
  }

  return missing;
}

console.log(missingLetter("the quick brown box jumps over a lazy dog"));

File: from-books/hash-table-exercises/language/rust/exercise-3.rs
use std::collections::HashSet;

fn generate_alphabet() -> HashSet<char> {
    let mut alphabet = HashSet::new();
    for i in b'a'..=b'z' {
        alphabet.insert(i as char);
    }
    alphabet
}

fn missing_letter(sentence: &str) -> Option<char> {
    let alphabet_set = generate_alphabet();
    let mut sentence_set = HashSet::new();

    for letter in sentence.to_lowercase().chars() {
        if letter.is_alphabetic() {
            sentence_set.insert(letter);
        }
    }

    alphabet_set.difference(&sentence_set).cloned().next()
}

// Write a function that acceps a string that contains all the letters of the alphabet
// except one and returns the missing letter. For example, the string "the quick brown box
// jumps over a lazy dog" contains all the letters of the alphabet except the letter, "f".
// The function should have a time complexity of O(n).

fn main() {
    match missing_letter("the quick brown box jumps over a lazy dog") {
        Some(missing) => println!("Missing letter: {}", missing),
        None => println!("No letters missing"),
    }
}
